### PR TITLE
The Day Has Come!

### PR DESCRIPTION
**Pull request by adamemerson**
Switch from C++14 to C++17 and remove backports.

### CODE REVIEW

**Comment by tchaikov** [comment]
this PR depends on #19645

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -157,7 +158,7 @@ if(CMAKE_VERSION VERSION_LESS "3.1")
   endif()
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
 else()
-  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: the C++17 support of `CMAKE_CXX_STANDARD` was introduced in cmake v3.8, see https://cmake.org/cmake/help/v3.8/prop_tgt/CXX_STANDARD.html and https://cmake.org/cmake/help/v3.7/prop_tgt/CXX_STANDARD.html, so we need to bump the required cmake version at https://github.com/ceph/ceph/pull/19657/files#diff-95e351a3805a1dafa85bf20b81d086e6R145 to `3.8`.

Comment (adamemerson) [reply src/CMakeLists.txt]: All right, got it updated.

**Comment by tchaikov** [comment]
retest this please.

**Comment by liewegas** [comment]
Just want to confirm that there are *no* effects of this transition on the C ABI or other changes that might affect downstream builds.... we need to be sure we aren't causing problems there before this merges

**Comment by adamemerson** [comment]
It shouldn't affect the C ABI at all.

We could try running a few test builds of things against it, maybe QEMU, if you think that might be a good test.

**Comment by adamemerson** [comment]
*C++* ABI might change because noexcept is part of the type and affects name mangling now.

(Though I'm not sure if that's an issue for us, since I think we break C++ ABI every time we touch bufferlist.)

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-01-04_11:59:52-rados-wip-kefu-testing-2018-01-04-1836-distro-basic-smithi/

failures tracked by http://tracker.ceph.com/issues/22527 and http://tracker.ceph.com/issues/22596

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-01-05_15:34:38-upgrade-wip-kefu-testing-2018-01-04-1836-distro-basic-mira/

the rbd upgrade test failed due to http://tracker.ceph.com/issues/22597

**Comment by tchaikov** [comment]
i think after switching to GCC-7, the ubuntu build of librados/librbd/libcephfs are not compatible with old client anymore. as they are using the new C++ ABI. which is not compatible with the old one.

**Comment by liewegas** [comment]
Should we bump the SONAME (finally)?

**Comment by cbodley** [comment]
@adamemerson these version numbers are in our cmake build, could you update them here? a `grep -r SOVERSION --include=CMakeLists.txt` should find them all, though i'm not sure whether this would be a new major or minor version

**Comment by adamemerson** [comment]
Will in a second, just have to finish this backport first.

**Comment by adamemerson** [comment]
After discussions with Sage and Greg, we'll not bump the SOVERSION. At some later date I'll split off libradospp from librados so they can be independently versioned, but right now bumping them would just break librados C compatibility.

**Comment by tchaikov** [comment]
@adamemerson have you started the split work? i can help on it if you like. the sooner we do the C++17 switch the merrier, I think.

**Comment by adamemerson** [comment]
I don't think the libradospp split needs to block this. As I understand it, the consensus is that so long as we aren't breaking the C ABI and we send a note to the list about the C++ ABI we'll be fine.

**Comment by tchaikov** [comment]
@adamemerson awesome! could you rebase this change against master? then it's good to merge then, i guess.

**Comment by adamemerson** [comment]
@tchaikov All right! We should be good to go now, I think.

File: src/include/scope_guard.h [hunk 2]
```diff
@@ -30,7 +30,7 @@ struct scope_guard {
   scope_guard(const F& f) : f(f) {}
   scope_guard(F &&f) : f(std::move(f)) {}
   template<typename... Args>
-  scope_guard(ceph::in_place_t, Args&& ...args) : f(std::forward<Args>(args)...) {}
+  scope_guard(std::in_place_t, Args&& ...args) : f(std::forward<Args>(args)...) {}
```
Comment (tchaikov) [inline src/include/scope_guard.h]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/include/scope_guard.h:31:30: error: expected ')' before ',' token
   scope_guard(std::in_place_t, Args&& ...args) : f(std::forward<Args>(args)...) {}
```

@adamemerson

**Comment by adamemerson** [comment]
That's odd, that line isn't in my source. Time to build commit by commit, I guess and see where it's finding it.

**Comment by adamemerson** [comment]
@tchaikov Can you look at this build failure? It claims to be a trusty+xenial pbuilder and it looks like it just has C++17 definitions missing. Could the wrong pbuilder have been turned on?

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by adamemerson** [comment]
Apparently the cmake version bump got lost when I rebased. +1 to @cbodley for spotting it.

Should work now.

Let us see.

**Pull request merged**